// minecraft-server-util/dist/statusBedrock.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.statusBedrock = void 0;
const assert_1 = __importDefault(require("assert"));
const crypto_1 = require("crypto");
const minecraft_motd_util_1 = require("minecraft-motd-util");
const UDPClient_1 = __importDefault(require("./structure/UDPClient"));
const srvRecord_1 = require("./util/srvRecord");
function statusBedrock(host, port = 19132, options) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, assert_1.default)(typeof host === 'string', `Expected 'host' to be a 'string', got '${typeof host}'`);
        (0, assert_1.default)(host.length > 1, `Expected 'host' to have a length greater than 0, got ${host.length}`);
        (0, assert_1.default)(typeof port === 'number', `Expected 'port' to be a 'number', got '${typeof port}'`);
        (0, assert_1.default)(Number.isInteger(port), `Expected 'port' to be an integer, got '${port}'`);
        (0, assert_1.default)(port >= 0, `Expected 'port' to be greater than or equal to 0, got '${port}'`);
        (0, assert_1.default)(port <= 65535, `Expected 'port' to be less than or equal to 65535, got '${port}'`);
        let srvRecord = null;
        if (typeof options === 'undefined' || typeof options.enableSRV === 'undefined' || options.enableSRV) {
            srvRecord = yield (0, srvRecord_1.resolveUDPSRV)(host);
            if (srvRecord) {
                host = srvRecord.host;
                port = srvRecord.port;
            }
        }
        const socket = new UDPClient_1.default(host, port);
        try {
            // Unconnected ping packet
            // https://wiki.vg/Raknet_Protocol#Unconnected_Ping
            {
                socket.writeByte(0x01);
                socket.writeInt64BE(BigInt(Date.now()));
                socket.writeBytes(Uint8Array.from([0x00, 0xff, 0xff, 0x00, 0xfe, 0xfe, 0xfe, 0xfe, 0xfd, 0xfd, 0xfd, 0xfd, 0x12, 0x34, 0x56, 0x78]));
                socket.writeBytes((0, crypto_1.randomBytes)(4));
                yield socket.flush(false);
            }
            // Unconnected pong packet
            // https://wiki.vg/Raknet_Protocol#Unconnected_Pong
            {
                const packetType = yield socket.readByte();
                if (packetType !== 0x1C)
                    throw new Error('Expected server to send packet type 0x1C, received ' + packetType);
                yield socket.readInt64BE();
                const serverGUID = yield socket.readInt64BE();
                yield socket.readBytes(16);
                const responseLength = yield socket.readInt16BE();
                const response = yield socket.readString(responseLength);
                const [edition, motdLine1, protocolVersion, version, onlinePlayers, maxPlayers, serverID, motdLine2, gameMode, gameModeID, portIPv4, portIPv6] = response.split(';');
                const motd = (0, minecraft_motd_util_1.parse)(motdLine1 + '\n' + motdLine2);
                return {
                    edition,
                    motd: {
                        raw: (0, minecraft_motd_util_1.format)(motd),
                        clean: (0, minecraft_motd_util_1.clean)(motd),
                        html: (0, minecraft_motd_util_1.toHTML)(motd)
                    },
                    version: {
                        name: version,
                        protocol: parseInt(protocolVersion)
                    },
                    players: {
                        online: parseInt(onlinePlayers),
                        max: parseInt(maxPlayers)
                    },
                    serverGUID,
                    serverID,
                    gameMode,
                    gameModeID: parseInt(gameModeID),
                    portIPv4: parseInt(portIPv4),
                    portIPv6: parseInt(portIPv6),
                    srvRecord
                };
            }
        }
        finally {
            socket.close();
        }
    });
}
exports.statusBedrock = statusBedrock;
